### ISSUE TITLE
emsdk 2.0.9 + VirtualBox shared folder = "parse exception"

### ISSUE THREAD

**Issue by hostilefork** [emscripten-core/emsdk#677]
**TL; DR:** Using **wasm-ld** on a trivial object file (e.g. the .o for `int main() { return 0; }`) will produce an output .wasm file of all 00 bytes if targeting a VirtualBox shared folder (Windows host, Linux guest).  This behavior started in emsdk 2.0.9.

Researching the commit reported in `wasm-ld --version`

Version that comes with 2.0.9 circa Nov 13 gives this (has the problem):
https://github.com/llvm/llvm-project/commit/27e9f0f95ef7b144d008bc1cf1459dced6cb5842

Version that comes with 2.0.8 circa Oct 24 (does not have the problem):
https://github.com/llvm/llvm-project/commit/1a7a9efec3cf872dcf3e1a6e6fe39e797c4d9fc6

It only spans a few commits to the `lld/wasm` directory.  But if I'm reading correctly, this spans [*2,124 commits*](https://github.com/llvm/llvm-project/compare/1a7a9efec3cf872dcf3e1a6e6fe39e797c4d9fc6...27e9f0f95ef7b144d008bc1cf1459dced6cb5842) to the LLVM project as a whole in less than a month.  :-(  I'm not exactly sure how many of these could affect wasm-ld.

How this manifests varies, but it all seems to relate to the generated wasm files being full of zeros.  A test file with just `int main() { return 0; }` gives this:

    ~/vbox-shared$ emcc test.c -o test.html
    [parse exception expected more elements in list]
    Fatal: error in parsing input
    emcc: error /home/hostilefork/emsdk/upstream/bin/wasm-emscripten-finalize --detect-features --minimize-wasm-changes --dyncalls-i64 test.wasm -o test.wasm failed (1)

That same file works fine if it's on the VM's Linux filesystem:

    ~$ emcc test.c -o test.html
    ~$

I reproduced the effect on two different computers (running similar configurations of VirtualBox 6.1.16).

This doesn't seem like the usual "using illegal file characters" kind of problem.  It may be worth narrowing down to find out which commit is causing it...and if it was specifically a wasm-oriented commit?  Perhaps those more in the know would have an idea.

**Comment by sbc100** [comment]
Very odd.

Have you verified that the wasm-ld is really writing all zeros?    If you build with `EMCC_DEBUG=1` then emscripten will save all its intermediate files in `/tmp/emscripten_temp` and also show you the internal wasm-ld command.

You should then be able to create a simple wasm-ld command that then fails without any of the emscripten stuff.    Once you have that there are a couple of ways to bisect.  

1.  Checkout llvm-project and use git bisect.   You can run something like `ninja lld && wasm-ld .... -o output.wasm && wasm-validate output.wasm`  as your bisect script (not the you only build lld at each step not all of llvm).
2.  Modify emsdk to install specific build.   The list of all emsdk builds is at https://ci.chromium.org/p/emscripten-releases/builders/ci/linux.   To install a given set of binaries you need to modify `emscripten-releases-tot.txt` in the emsdk directory to contain the tag you want and then run `emsdk install tot`.   That should give you much smaller range of llvm commits.

**Comment by hostilefork** [comment]
> You should then be able to create a simple wasm-ld command that then fails without any of the emscripten stuff.

From comparing the EMCC_DEBUG output, it seems wasm-ld can't write its output to virtualbox shared folders in any (recent) version. 

So what's *actually* causing the change is that 2.0.8 uses a temporary file location for its output, while 2.0.9 uses the current directory.  e.g. 2.0.8 has the line:

    /home/hostilefork/emsdk/upstream/bin/wasm-ld -o /tmp/emscripten_temp/test.wasm /tmp/emscripten_temp/test_0.o ...

Then in 2.0.9 it uses the current directory--in my case, a virtualbox shared folder, which makes the zeros for some reason:

    /home/hostilefork/emsdk/upstream/wasm-ld -o test.wasm /tmp/emscripten_temp/test_0.o ...

If that change was done deliberately for a meaningful reason, then...oh well.  But if the change wasn't on purpose, then building to a temporary file has at least one advantage in not triggering this odd lld quirk.  :-/

Following your advice I did manage to build lld for wasm.  So if I can find a version that doesn't make corrupt files on VirtualBox in the history, perhaps I can bisect it and locate the problem.  Or just debug it.  Notes on building just lld, just with wasm:

    $ git clone https://github.com/llvm/llvm-project --depth 1
    $ cd llvm-project
    llvm-project$ mkdir build
    llvm-project$ cd build
    llvm-project/build$ cmake -G Ninja -DLLVM_ENABLE_PROJECTS='lld' DLLVM_TARGETS_TO_BUILD='WebAssembly' -DCMAKE_BUILD_TYPE=Release
    llvm-project/build$ ninja lld

**Comment by sbc100** [comment]
OK so it sounds like its not an lld change that triggered this but instead the emscripten change that avoids writing the temporary directory where possible.  

So this command fails silently when the CWD is a vistualbox shared directory: `wasm-ld -o test.wasm /tmp/emscripten_temp/test_0.o` but its works fine when the CWD is a normal linux directory?    Sounds like make a virtualbox bug?   Does normal `lld` or GNU ld exhibit similar problems?   (I don't think wasm-ld differs from normal (ELF) lld in how it writes the final output)

**Comment by hostilefork** [comment]
> Does normal lld or GNU ld exhibit similar problems?

GNU ld does not seem to have a problem:

    ~/vboxsf$ as hello.as -o hello.o  # assembler file with _start()

    ~/vboxsf$ ld hello.o -o hello
    ~/vboxsf$ ./hello
    Hello, world

But lld does seem to (I used **sudo apt install lld**, as it didn't seem to come with clang).

    ~/vboxsf$ lld
    lld is a generic driver.
    Invoke ld.lld (Unix), ld64.lld (macOS), lld-link (Windows), wasm-ld (WebAssembly) instead

    ~/vboxsf$ ld.lld hello.o -o hello
    ~/vboxsf$ ./hello
    bash: ./hello: cannot execute binary file: Exec format error
    -rwxrwxrwx 1 hostilefork root 1080 Dec 31 13:02 hello*

    ~/vboxsf$ xxd hello
    00000000: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    00000010: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    00000020: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    (all zeros in file)

But it works in a local directory, and it works if you use clang on a C object:

    ~/vboxsf$ clang -c hello.c  # C file with int main() and printf() hello world

    ~/vboxsf$ clang -o hello hello.o    
    ~/vboxsf$ ./hello
    Hello, world

So clang must have some different method of writing files (?)

> Sounds like a virtualbox bug?

Probably...though I've not had any other trouble that was data corruption oriented like this.  Just issues with not supporting symbolic links and some filenames.

But if your change to not use a temporary file was intended, then I guess it's best to close this as not emscripten's problem.

**Comment by sbc100** [comment]
It was an intentional change yes.   But it still very strange and probably worth investigating.     The wasm file should obviously never contains all zeros no matter where its written.

Can you confirm if running wasm-ld on its own produces a file containing all zeros?

**Comment by hostilefork** [comment]
>  But it still very strange and probably worth investigating.

On a [page for **`llvm::raw_ostream`**](https://llvm.org/doxygen/classllvm_1_1raw__ostream.html) it describes it as *"an extremely fast bulk output stream that can only output to a stream"*.  That seems a bit of a smoking gun, unless clang uses it too (?)

Maybe a test could be pared down that makes such a stream, writes to it, and closes it.

> Can you confirm if running wasm-ld on its own produces a file containing all zeros?

It acts the same as the scenario given above for ld.lld.

**Comment by sbc100** [comment]
Oh I see.. I didn't see that `ld.lld hello.o -o hello` produces the problem.   In that case can file a bug in upstream llvm ( https://bugs.llvm.org/)?   This seems to be an lld issue and not am emscripten one.   You can link back o this issue and keep his one open.

**Comment by sbc100** [comment]
I believe this is an issue with writing the windows filesystem from the linux VM guest machine.  I don't think its specific to, or related to, emsdk so closing this out.
